### ISSUE TITLE
Theme TooWaBlue updates

### ISSUE THREAD

**Issue by TooWaBoo** [Beep6581/RawTherapee#3214]
Update v1.17, 2016-04-03 - Needs branch **more-object-id**
[TooWaBlue.css.txt](https://github.com/Beep6581/RawTherapee/files/201573/TooWaBlue.css.txt)

Deutsch locale update - small corrections
[Deutsch.txt](https://github.com/Beep6581/RawTherapee/files/201575/Deutsch.txt)

**Comment by Hombre57** [comment]
Could you please use only one issue for your updates ?

**Comment by TooWaBoo** [comment]
Sorry, I'll do it next time.
You can close #3211 and I will take this one for the future.

**Comment by TooWaBoo** [comment]
Update v1.13 uploaded in first post.

**Comment by Beep6581** [comment]
Committed.

**Comment by TooWaBoo** [comment]
Update v1.14 uploaded in first post.

**Comment by Beep6581** [comment]
Hey

`GtkPaned:nth-child(1) > GtkBox > GtkGrid:nth-child(1) > GtkContainer > GtkDrawingArea:nth-child(1)`
That I agree is quite horrible and will break with the slightest breeze. Adding classes or IDs is the way to go, though not always possible. I can help to some degree, but for some of RT's custom widgets we will need @Hombre57's help.

Could you clarify exactly what it is you want the ID in each of those things in the list applied to? For example, when you say Queue do you mean the frame or notebook that holds it? What Partial Paste widget? And so forth. Speaking of Partial Paste, why do you want a unique ID for it?

**Comment by TooWaBoo** [comment]
To add a frame to the partial paste dialog I do this:
```
GtkDialog > GtkBox:nth-child(1) > GtkScrolledWindow {
	background-color: @bg-light-grey;
	border: 1px solid @bg-button-border;
}
```
This is unique now. There is no dialog with the same object tree.
If it's possible to set an ID for "GtkDialog" or "GtkBox" I can use:
```
#PartialPaste GtkScrolledWindow
```
---------
For e.g. the Histogram with an assigned ID to GtkBox or higher parent, I can use:
```
#Histogram GtkDrawingArea:nth-child(1) -> Histogramm
#Histogram GtkDrawingArea:nth-child(2) -> RGB-Bar
```
--------
I'm referencing the button in the "Queue" with this selector:
```
#MainNotebook > GtkBox > GtkBox:nth-child(1) > GtkBox:nth-child(1) .button
```
The ID should be set on "GtkBox". Referencing the button can be done by:
```
#Queue .button
```
-----------------
The IDs should be set to the top most object. Tthe long selectors I'm using now are only necessary to make the reference unique.

**Comment by TooWaBoo** [comment]
I think, it's not so difficult.  I guess the code has to be changed like this:

1. batchqueuepanel.cc (an ID for the top button box would be enough)
```
...
BatchQueuePanel::BatchQueuePanel (FileCatalog* aFileCatalog)
{
...
    // setup button bar
    topBox = Gtk::manage (new Gtk::HBox ());
	
->   topBox->set_name("QueueButtonBox"); <-

    pack_start (*topBox, Gtk::PACK_SHRINK);
...
```
2. navigator.cc
```
...
Navigator::Navigator ()
{

    set_label (M("MAIN_MSG_NAVIGATOR"));
    Gtk::VBox* mbox = Gtk::manage (new Gtk::VBox ());
    previewWindow = Gtk::manage (new PreviewWindow ());

->   mbox->set_name("Navigator"); <-

    mbox->pack_start (*previewWindow, Gtk::PACK_SHRINK, 2);
    position = Gtk::manage (new Gtk::Label ());
	mbox->pack_start (*position, Gtk::PACK_SHRINK, 2);
...
```

3. histogrampanel.cc
```
...
HistogramPanel::HistogramPanel ()
{
...
    histogramRGBArea->set_halign(Gtk::ALIGN_FILL);
    histogramRGBArea->set_valign(Gtk::ALIGN_END);
    histogramRGBArea->show();
    gfxGrid = Gtk::manage (new Gtk::Grid ());
	
->   gfxGrid->set_name("Histogram"); <-
	
    gfxGrid->set_orientation(Gtk::ORIENTATION_VERTICAL);
    gfxGrid->set_row_spacing(1);
    gfxGrid->set_column_spacing(1);
...
```
4. partialpastedlg.cc
```
...
PartialPasteDlg::PartialPasteDlg (const Glib::ustring& title, Gtk::Window* parent)
    : Gtk::Dialog (title, *parent, true)
{
...
    Gtk::HBox* hbmain = Gtk::manage (new Gtk::HBox ());
	
->   hbmain->set_name("PartialPaste"); <-
	
    hbmain->pack_start (*vbCol1);
    hbmain->pack_start (*(Gtk::manage (new Gtk::VSeparator ())));
...
```
5. no clue :-)

**Comment by Hombre57** [comment]
5- `editorpanel.cc` line 354 :
```
    beforeAfterBox = Gtk::manage (new Gtk::HBox());
    beforeAfterBox->pack_start (*afterBox);

    beforeAfterBox->set_name("BeforeAfterContainer");

    editbox->pack_start (*toolBarPanel, Gtk::PACK_SHRINK, 2);
    editbox->pack_start (*beforeAfterBox);
```

I'm reluctant to call it `BeforeAfterBox` because it may be swapped to a Gtk::Grid in the future (Gtk::Box may be deprecated one day).

**Comment by TooWaBoo** [comment]
@Hombre57 
So it's possible to get these IDs, nice. Thanks
@Beep6581 
You should add version 15 to the branch. It contains some fixes. The long selectors make it save that only the specified objects are changed.

**Comment by Hombre57** [comment]
@TooWaBoo Why do you want naming top-level objects if you want to theme it's child ? It seems more logical to set an ID for custom `Gtk::DrawingArea` objects like `HistogramArea` and `HistogramRGBArea` than it's parent IMHO, because if we create a new instance of e.g. `HistogramArea` elsewhere, it won't be correctly skinned.

Do you have a list of specific objects that you want to set an ID to ?

**Comment by TooWaBoo** [comment]
@Hombre57 
There are always two sides
This is what I'm doing with the Histogram:
```
/*** Histogram ***/
GtkPaned:nth-child(1) > GtkBox > GtkGrid:nth-child(1) > GtkContainer > GtkDrawingArea:nth-child(1),
GtkPaned:nth-child(1) > GtkBox > GtkGrid:nth-child(1) > GtkContainer {
    ...<some styles>...
}
/*** Histogram RGB-Bar***/
GtkPaned:nth-child(1) > GtkBox > GtkGrid:nth-child(1) > GtkContainer > GtkDrawingArea:nth-child(2) {
    ...<some styles>...
}
```
With an ID of the top-level object, it will look like this:
```
/*** Histogram ***/
#Histogram GtkDrawingArea:nth-child(1),
#Histogram  {
    ...<some styles>...
}
/*** Histogram RGB-Bar***/
#Histogram GtkDrawingArea:nth-child(2) {
    ...<some styles>...
}
```
If you create another Histogram with the same look, you only need to set the same top-level ID.
It follows the same rule like in the right toolbar. Here's an example from the org. theme.
```
/* Tool background */
#ExpanderBox {
    ...<some styles>...
}

#ExpanderBox GtkDrawingArea {
    ...<some styles>...
}

#ExpanderBox GtkFrame {
    ...<some styles>...
}

#ExpanderBox GtkFrame GtkDrawingArea {
    ...<some styles>...
}

#ExpanderBox GtkFrame GtkFrame {
    ...<some styles>...
}

#ExpanderBox GtkFrame GtkFrame GtkDrawingArea {
    ...<some styles>...
}
```
The top-level object is used to apply the styles only to a specific area, in my case to a specific widget. If you have to create an ID for every individual object I'm accessing, I guess, you will kill me :-)

Right now, I need these long selectors only to limit the styles to a specific widget. Thats why I'm asking for IDs to the top-level objects.

**Comment by Hombre57** [comment]
See the new-object-id branch. I hope it will be self explaining, otherwise I'll explain the new IDs.

**Comment by TooWaBoo** [comment]
@Hombre57 
Thanks a lot. It's self explaning.

**Comment by Kildor** [comment]
Is using radiobutton sprite for checkboxes intended?
Blue dot instead of proper checkmark is really annoying and unobvious.
![image](https://cloud.githubusercontent.com/assets/1320579/14061786/14c74e3c-f3b4-11e5-8b43-1232d2cb38da.png)

RT Branch: gtk3
Version: 4.2.880
Changeset: 4979c0c83604cc489534b39f04bca9199d2ed6c9

Downloaded from the site

**Comment by TooWaBoo** [comment]
@Kildor 
I've removed all the military green buttons and check/radio images. The target was to build a theme which will not touch any resource outside the .css file and this was the best I could do.

Now the theme is part of RT, and I guess it's possible to replace the checkbox with an image, but for now the prioryty is low to do this. Maybe in the future.

**Comment by TooWaBoo** [comment]
@Hombre57 
The changes are done.
Can you provide a compiled version of the " new-object-id" branch for testing? A couple of weeks ago, I've tried to complie myself, but with no luck.
Thanks

**Comment by Hombre57** [comment]
I can, but for Windows 64 bits. Is it ok ?

**Comment by TooWaBoo** [comment]
@Hombre57 
Yes, that's what I need.
Thanks

**Comment by Hombre57** [comment]
Release build (only *.exe) available here : http://filebin.net/ao1lb7g7sv

**Comment by TooWaBoo** [comment]
@Hombre57 
I've copied the .exe into my install folder and received folling error message:
Something wasn't found.
![unbenannt](https://cloud.githubusercontent.com/assets/13800920/14066174/5156dbc6-f443-11e5-8133-c0102be36fb5.png)

**Comment by Hombre57** [comment]
Here is the full archive (no installer, just unzip to the dest location). http://filebin.net/ao1lb7g7sv/RawTherapee4_2_897-full.zip

**Comment by TooWaBoo** [comment]
@Hombre57 
Tried full version, RT quits immediately without any message.

**Comment by Hombre57** [comment]
So ask @sguyader for a build.

**Comment by TooWaBoo** [comment]
@Hombre57 
The new ID's work fine, except the #PartialPaste ID. hbmain doesn't support padding and will overlay the scrollbars if I set a background. Is it possible to set the #PartialPaste ID to the **scrolledwindow** ? Thats the only object to set the background color and the included child GtkViewPort supports padding.

```
Gtk::HBox* hbmain = Gtk::manage (new Gtk::HBox ());
    hbmain->set_name("PartialPaste");
    hbmain->pack_start (*vbCol1);
    hbmain->pack_start (*(Gtk::manage (new Gtk::VSeparator ())));
    hbmain->pack_start (*vbCol2);
    hbmain->pack_start (*(Gtk::manage (new Gtk::VSeparator ())));
    hbmain->pack_start (*vbCol3);

    scrolledwindow = Gtk::manage ( new Gtk::ScrolledWindow() );
    scrolledwindow->set_can_focus(true);
    scrolledwindow->set_shadow_type(Gtk::SHADOW_NONE);
    scrolledwindow->set_policy(Gtk::POLICY_AUTOMATIC, Gtk::POLICY_AUTOMATIC);
scrolledwindow->property_window_placement().set_value(Gtk::CORNER_TOP_LEFT);
```

I've also tried to change the checkbox to an image.
On Windows I can use...
```-gtk-icon-source: url("./TooWaBlueImages/....png");```
...to read the image from a folder "TooWaBlueImages" inside the theme folder.

Can anybody test if this will work on Linux too?
Here's a test theme.
http://filebin.net/m30tu9nadf
It uses the new ID's, so if you are not using the build from the "more object ID" some backgrounds looks wrong.

This is how it looks like:
![unbenannt](https://cloud.githubusercontent.com/assets/13800920/14139669/b9e16462-f674-11e5-98cf-b01fc9741077.png)

**Comment by TooWaBoo** [comment]
ping :-)

**Comment by Hombre57** [comment]
Modification committed to the more-object-id branch.

@TooWaBoo 
- Did you get a build from @sguyader ?
- Could you add a checkbox image for the *inconsistent* state ? (see the Batch tool panel when editing several images with different values for the same parameter).

I think it's better to add the images as a subfolder in theme, as it is specific to your theme. It would be nice if a linux user could test the new temporary theme.

**Comment by TooWaBoo** [comment]
@Hombre57 
Yes, I got a build from @sguyader
The image for the inconsistent checkbox is already done but I forgot the radio button. I'll do it now.

**Comment by adamreichold** [comment]
@TooWaBoo @Hombre57 Tested on Linux using Gtk+ 3.18.9 and it looks more or less like the above screenshot from Windows:
![color_management_linux](https://cloud.githubusercontent.com/assets/12997846/14231385/8a35f51a-f981-11e5-9675-f0529c0d6668.png)

**Comment by TooWaBoo** [comment]
@adamreichold 
Thanks for testing. So I keep the images inside the theme folder.
Now I'm waiting for @sguyader for the test build of the partial paste.

**Comment by Beep6581** [comment]
Hello

Branch `new-object-id` seems to work fine in Gentoo using Gtk+ 3.18.7.
The progress bar by the Preferences button is not visible in the default state, its only visible when it's being used. Maybe this is intentional, I just wanted to point it out.

@TooWaBoo to conform better to RT style, could you introduce these changes?
1. Indent using spaces not tabs. 1 tab = 4 spaces.
2. When merging, delete `TooWaBlue.css` and rename `TooWaBlueCheckbox.css` to `TooWaBlue.css`
3. Provide SVG versions of the images.
4. Image folder structure:
    PNG goes here: `rtdata/themes/images/twb/`, no need for "TWB-" in each image filename.
    SVG goes here: `rtdata/themes/images/svg/twb/`
    If we need to batch render PNG images in a different resolution or color, this will make doing so easier.

**Comment by TooWaBoo** [comment]
@Hombre57 
PartialPaste works. (Thanks to @sguyader for compiling).

@Beep6581 
The behaviour of the top progressbar is intentionel.

1. done

2. The TooWaBlueCheckbox.css should not be in any branch. It was only for testing if the relativ path I use also works in Linux.
I didn't add the .css file to the branch, I've no permissions to do it and I don't have any clue how to do it. :-)

3. Checkbox images are treated as icons. SVG will not work.

4. If I understand you right, then my images are in a folder "twb" inside the "images" folder.
In Windows I can access the folder "images" from a .css file by the following path:
```../images/twb/ImageName.png ```
The two dots in front of the path points to the parent folder of the .css file location and there has to be the folder "images". Does this work in Linux too? If you want, I can provide a test theme..

**Comment by Beep6581** [comment]
> Checkbox images are treated as icons. SVG will not work.

Yes, but we need the SVG files so that we can re-generate the PNG files when needed, e.g. when changing a color, or when adding support for hi-dpi, so please provide the SVG files too.
Also, Gtk+ 3 supports SVG as far as I know, so if we got it working for RT that would solve some problems down the line, such as hi-dpi support.

> The two dots in front of the path points to the parent folder of the .css file location and there has to be the folder "images". Does this work in Linux too? 
What works in Windows should work in Linux because this is supposed to follow the CSS standard (to some degree). I can test once point 4 is done.

**Comment by TooWaBoo** [comment]
@Beep6581 
Hmmm, I don't have the images as vector and I don't have an application to create vector gaphics . It's pixel work. When I converting them in SVG, it's only a container with a PNG included. No way to get a higher resolution or changing colors.

**Comment by TooWaBoo** [comment]
I've decided to not using images and went back to the gradients with the round checkmark. There are no issues for higher resolution.
The inconsistent status will be indicated bei a blue dot (see screenshot).
The new version need the **"more-object-ID"** branch to work properly.
**The new version is in the first post.**

![unbenannt](https://cloud.githubusercontent.com/assets/13800920/14235039/302a2cb8-f9f3-11e5-9dcd-a7d5919a653c.PNG)

**Comment by Kildor** [comment]
It is very bad. Dots can`t represent checkboxes.
It was good theme, I will miss it.

**Comment by Hombre57** [comment]
@Kildor I agree that checkmark is better here.

@TooWaBoo Building svg from the png files shouldn't be too complex. I'll do it tonight, so please keep the checkmark version.

**Comment by TooWaBoo** [comment]
@Hombre57 
Here you will find all the latest files for converting and testing.
http://filebin.net/d2hdijyniq
**TooWaBlue.css** v1.19
**twb** folder with the images
only **exe** of RT from the "more-object-id" branch compiled by @sguyader

The folder **twb** has to be in the folder **images**

**Comment by Beep6581** [comment]
I agree, the checkmarks look good. I thought you started off from SVG. Don't worry, we'll help.

**Comment by Beep6581** [comment]
Here they are:
http://filebin.net/gzn4u24kmm
SVG and PNG, light and dark. Please just arrange them in a folder structure as described in the comment above https://github.com/Beep6581/RawTherapee/issues/3214#issuecomment-204973231

**Comment by TooWaBoo** [comment]
@Beep6581 
Thanks for the work, but one image is wrong: **checkbox-inconsistent-d.png**
The darker images are not dark and are not transparent. They have the same color as the insensitive font. The file name also ends with a **-d** (for disabled).
![unbenannt](https://cloud.githubusercontent.com/assets/13800920/14254511/57fb2d12-fa90-11e5-868a-093805ce3091.png)

**Comment by Beep6581** [comment]
That's the problem with calling things "`-d`", we must guess what it means. Let's not do that, and let's use full words instead.

We use a script to generate icons from SVG, the script automatically places the icons in designated folders. The images in the "dark" folder are always light, because the "dark" in the folder name refers to the background they are meant to be shown over. In your theme's case the background is always the same, so the script's way of naming things isn't useful here, that's why I asked that the files be renamed and sorted manually as described a few comments above.

The PNG icons I made the SVG from did not have a "square in a square" icon as in your screenshot, they had a dash as in my SVG. If that is incorrect, please send me a ZIP with the correct PNG files, then I will update the SVGs.

**Comment by TooWaBoo** [comment]
It's all there: https://github.com/Beep6581/RawTherapee/issues/3214 :-)

So I have to change the names to e.g **checkbox-checked-active.png**, **checkbox-checked-inactive.png**? If "Yes" I need to change the TooWaBlue.css too. - Thats not a problem.

The opacity is a problem. The checkbox will be shown on a light grey and a normal grey background, but the font brightness is always the same. If the checkbox is transparent they will not suit the font color
![b9e16462-f674-11e5-98cf-b01fc9741077](https://cloud.githubusercontent.com/assets/13800920/14255873/8ca641a4-fa96-11e5-9bfd-e0f382fa4f0b.png)
![unbenannt](https://cloud.githubusercontent.com/assets/13800920/14255984/2ffaff48-fa97-11e5-8991-5242caf65f5f.PNG)

**Comment by TooWaBoo** [comment]
Changed image names and css.

Here are the latest files for converting and testing.
http://filebin.net/7f2yglqcc6
**TooWaBlue.css v1.19.1**
**twb** folder with the images
**exe only** of RT from the "more-object-id" branch compiled by @sguyader

The folder **twb** has to be in the folder **images**

**Comment by Beep6581** [comment]
I'm working on it and hope to offer a PR soon.

**Comment by Beep6581** [comment]
RawTherapee uses a script for generating PNG icons from SVG. This script always generates a light-background and dark-background variant of the SVG icons. Generating icons for a custom theme is not handled optimally by the script because in the case of a custom theme the background never changes.

Because of this, after processing your theme SVG icons I used the dark-background version of the for the normal icon states, and the light-background version for the `-disabled` icons states. I think it worked well. The icons seem to match the font color just fine.
![scrot_2016-04-04 215247](https://cloud.githubusercontent.com/assets/5844619/14261331/64e65cb6-fab0-11e5-92f6-c06f9f162a02.png)
![scrot_2016-04-04 215507](https://cloud.githubusercontent.com/assets/5844619/14261330/64e478b0-fab0-11e5-83b0-526c75005ef6.png)

If that's ok with you, I'd like to commit. If not, it means we will have to manually save these icons whenever there is a change, and I would like to avoid that, so I hope your taste can adapt to what you see in the screenshot :]

**Comment by TooWaBoo** [comment]
Thanks, thats OK for me

**Comment by Beep6581** [comment]
Yay! @Hombre57 ok to merge?

**Comment by Hombre57** [comment]
@Beep6581 I'm quite busy, so if you don't mind, I'll let you do, and delete the more-object-id branch. I've nothing to add, all that @TooWaBoo requested has been done.

**Comment by Beep6581** [comment]
Done.

**Comment by TooWaBoo** [comment]
Installed the new RT 4.2.915 and the disabled checkbox and radio button are to dark.
I've modified the included SVG by hand to fit font color/brightness and I changed the stroke-width for the square of the checkbox to make it look a bit smoother. THE PNG were created by an online convertertool from SVG to PNG.

The opacity is set 70%, but what I did is to double the vector image.
opacity0.7 + opacity0.7 = opacity1.4 -> opacity = 1 :-)

RT is using colored images, so to set a specific color can't be a problem. I changed the color inside the SVG to the grey to fit the font color/brightness.

You can download the PNG and SVG here: http://filebin.net/ave0u5otci

Left TWB, Right Beep
![twb](https://cloud.githubusercontent.com/assets/13800920/14324519/698c700e-fc25-11e5-8e18-f634d95109a9.png)
